src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleNull.cs

```diff
@@ -16,8 +16,8 @@ private static bool HandleNull(ref Utf8JsonReader reader, ref ReadStack state, J
                 return false;
             }
 
-            // If we don't have a valid property, that means we read "null" for a root object so just return.
-            if (state.Current.JsonPropertyInfo == null)
+            // If null is read at the top level and the type is nullable, then the root is "null" so just return.
+            if (reader.CurrentDepth == 0 && (state.Current.JsonPropertyInfo == null || state.Current.JsonPropertyInfo.CanBeNull))
```

Comment:
Since this code flow ultimately results in a return from `ReadCore`, it should only ever be triggered at the top level (`reader.CurrentDepth == 0`)
Second condition is to ensure that the type we're dealing with is nullable.

Reply:
There is a helper called `IsLastFrame` on the stack frame that could be used instead of `CurrentDepth` but otherwise looks correct.

Reply:
`IsLastFrame` behaves slightly different from `reader.CurrentDepth`.
For example, when the reader encounters `null` in the following json `{ "MyProperty": null }`, `reader.CurrentDepth` is 1, but `IsLastFrame` is `true`.

`IsLastFrame` appears to only update when a new frame is pushed onto the stack, which happens while parsing sub-objects/sub-arrays within an object.